compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3854,6 +3854,176 @@ static void arraySetDefault(TR::Node* node, uint8_t elementSize, TR::Register* a
+   generateMemRegInstruction(TR::InstOpCode::MOVUPSMemReg, node, generateX86MemoryReference(addressReg, sizeReg, 0, -16, cg), xmmValueReg, cg);
+   }
+
+// Assumes valueReg is an XMM register that contains a vector of elements, and that sizeReg is within the range 32..63.
+static void arraySet32to63Bytes(TR::Node* node, uint8_t elementSize, TR::Register* addressReg, TR::Register* xmmValueReg, TR::Register* sizeReg, TR::CodeGenerator* cg)
+   {
+   generateMemRegInstruction(TR::InstOpCode::MOVUPSMemReg, node, generateX86MemoryReference(addressReg, 0, cg), xmmValueReg, cg);
+   generateMemRegInstruction(TR::InstOpCode::MOVUPSMemReg, node, generateX86MemoryReference(addressReg, 16, cg), xmmValueReg, cg);
+   generateMemRegInstruction(TR::InstOpCode::MOVUPSMemReg, node, generateX86MemoryReference(addressReg, sizeReg, 0, -32, cg), xmmValueReg, cg);
+   generateMemRegInstruction(TR::InstOpCode::MOVUPSMemReg, node, generateX86MemoryReference(addressReg, sizeReg, 0, -16, cg), xmmValueReg, cg);
+   }
+
+// Assumes valueReg is an XMM register that contains a vector of elements, and that sizeReg is >= 64.
+static void arraySet64ByteLoop(TR::Node* node, uint8_t elementSize, TR::Register* addressReg, TR::Register* xmmValueReg, TR::Register* sizeReg, TR::Register *scratch1Reg, TR::Register *scratch2Reg, TR::CodeGenerator* cg)
+   {
+   TR::LabelSymbol *loopLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *residueLabel = generateLabelSymbol(cg);
+
+   // Unaligned store to the first 16 bytes
+   generateMemRegInstruction(TR::InstOpCode::MOVUPSMemReg, node, generateX86MemoryReference(addressReg, 0, cg), xmmValueReg, cg);
+   generateRegRegInstruction(TR::InstOpCode::MOVRegReg(), node, scratch1Reg, addressReg, cg);
+   // Advance the address reg to the next 16-byte aligned address
+   generateRegImmInstruction(TR::InstOpCode::ADDRegImms(), node, addressReg, 15, cg);
+   generateRegImmInstruction(TR::InstOpCode::ANDRegImms(), node, addressReg, -16, cg);
+   // Calculate how many unaligned bytes were stored and subtract from the size
+   generateRegRegInstruction(TR::InstOpCode::SUBRegReg(), node, scratch1Reg, addressReg, cg);
+   generateRegRegInstruction(TR::InstOpCode::ADDRegReg(), node, sizeReg, scratch1Reg, cg);
+   // Check if we have at least 64 bytes left to store
+   // Enter the 64-byte per iteration aligned store loop if so
+   // Otherwise go to the residue loop
+   generateRegImmInstruction(TR::InstOpCode::CMPRegImm4(), node, sizeReg, 64, cg);
+   generateLabelInstruction(TR::InstOpCode::JB4, node, residueLabel, cg);
+   generateLabelInstruction(TR::InstOpCode::label, node, loopLabel, cg);
+   // 64-byte per iteration aligned store loop
+   generateMemRegInstruction(TR::InstOpCode::MOVAPSMemReg, node, generateX86MemoryReference(addressReg, 0, cg), xmmValueReg, cg);
+   generateMemRegInstruction(TR::InstOpCode::MOVAPSMemReg, node, generateX86MemoryReference(addressReg, 16, cg), xmmValueReg, cg);
```

Comment:
movaps requires alignment (to the size of the vector length), is the memory reference always aligned?